File: android/src/main/java/com/hlabexmaples/easyapi/ui/base/BaseFragment.java
package com.hlabexmaples.easyapi.ui.base;

import android.arch.lifecycle.MutableLiveData;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.hlab.easyapi_java.easyapi.util.LoaderInterface;
import com.hlab.easyapi_java.easyapi.util.STATE;
import com.hlabexmaples.easyapi.R;
import com.hlabexmaples.easyapi.common.DialogUtils;
import java.util.Objects;

/**
 * Base class for all the fragments used, manages common feature needed in the most of the fragments
 */
public abstract class BaseFragment<T extends ViewDataBinding> extends Fragment
    implements View.OnClickListener, LoaderInterface {

  protected T binding;

  protected abstract int defineLayoutResource();

  protected abstract void initializeComponent(View view);//to initialize the fragments components

  protected abstract void initToolbar();

  @Override
  public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    binding = DataBindingUtil.inflate(inflater, defineLayoutResource(), container, false);
    return binding.getRoot();
  }

  @Override
  public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    initToolbar();
    initializeComponent(view);
  }

  public void goBack() {
    if (getActivity() != null) {
      getActivity().onBackPressed();
    }
  }

  @Override
  public void showNoInternet() {
    DialogUtils.displayDialog(getActivity(), getString(R.string.alert_no_connection));
  }

  @Override
  public void showLoading() {

  }

  @Override
  public void hideLoading() {

  }

  @Override
  public void showError(String message) {
    DialogUtils.displayDialog(getActivity(), message);
  }

  public void observeLoadingState(MutableLiveData<STATE> state) {
    if (state != null) {
      state.observe(this, state1 -> {
        switch (Objects.requireNonNull(state1)) {
          case SHOW_LOADING:
            showLoading();
            break;
          case HIDE_LOADING:
            hideLoading();
            break;
          case ERROR:
            showError("");
            break;
          case NO_INTERNET:
            showNoInternet();
            break;
        }
      });
    }
  }

  @Override
  public void onHiddenChanged(boolean hidden) {
    super.onHiddenChanged(hidden);
    if (!hidden) {
      initToolbar();
    }
  }

  /**
   * Adds the Fragment into layout container.
   *
   * @param container Resource id of the layout in which Fragment will be added
   * @param currentFragment Current loaded Fragment to be hide
   * @param nextFragment New Fragment to be loaded into container
   */
  protected void add(final int container, final Fragment currentFragment,
      final Fragment nextFragment) {
    if (getFragmentManager() != null) {
      getFragmentManager().beginTransaction()
          .add(container, nextFragment, nextFragment.getClass().getSimpleName())
          .hide(currentFragment)
          .addToBackStack(null)
          .commit();
    }
  }
}

File: kotlin/src/main/java/com/hlabexamples/easyapi/kotlin/ui/base/BaseFragment.kt
package com.hlabexamples.easyapi.kotlin.ui.base

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Observer
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.hlab.easyapi_kotlin.easyapi.util.LoaderInterface
import com.hlab.easyapi_kotlin.easyapi.util.STATE
import com.hlabexamples.easyapi.kotlin.common.DialogUtils
import com.hlabexmaples.easyapi.R

/**
 * Base class for all the fragments used, manages common feature needed in the most of the fragments
 */
abstract class BaseFragment<T : ViewDataBinding> : Fragment(),
    View.OnClickListener,
    LoaderInterface {

  protected lateinit var binding: T

  protected abstract fun defineLayoutResource(): Int

  protected abstract fun initializeComponent(view: View) //to initialize the fragments components

  protected abstract fun initToolbar()

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    binding = DataBindingUtil.inflate(inflater, defineLayoutResource(), container, false)
    return binding.root
  }

  override fun onViewCreated(
    view: View,
    savedInstanceState: Bundle?
  ) {
    super.onViewCreated(view, savedInstanceState)
    initToolbar()
    initializeComponent(view)
  }

  fun goBack() {
    if (activity != null) {
      activity!!.onBackPressed()
    }
  }

  override fun showNoInternet() {
    DialogUtils.displayDialog(context!!, getString(R.string.alert_no_connection))
  }

  override fun showLoading() {

  }

  override fun hideLoading() {

  }

  override fun showError(message: String) {
    DialogUtils.displayDialog(context!!, message)
  }

  fun observeLoadingState(state: MutableLiveData<STATE>?) {
    state?.observe(this, Observer<STATE> { state1 ->
      state1?.let {
        when (it) {
          STATE.SHOW_LOADING -> showLoading()
          STATE.HIDE_LOADING -> hideLoading()
          STATE.ERROR -> showError("")
          STATE.NO_INTERNET -> showNoInternet()
        }
      }
    })
  }

  override fun onHiddenChanged(hidden: Boolean) {
    super.onHiddenChanged(hidden)
    if (!hidden)
      initToolbar()
  }

  /**
   * Adds the Fragment into layout container.
   *
   * @param container       Resource id of the layout in which Fragment will be added
   * @param currentFragment Current loaded Fragment to be hide
   * @param nextFragment    New Fragment to be loaded into container
   */
  protected fun add(
    container: Int,
    currentFragment: Fragment,
    nextFragment: Fragment
  ) {
    if (fragmentManager != null) {
      fragmentManager!!.beginTransaction()
          .add(container, nextFragment, nextFragment.javaClass.simpleName)
          .hide(currentFragment)
          .addToBackStack(null)
          .commit()
    }
  }
}